org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java

```diff
@@ -650,7 +670,7 @@ private CompletableFuture<LanguageServer> connect(@NonNull URI uri, IDocument do
 				DocumentContentSynchronizer listener = new DocumentContentSynchronizer(this, theDocument, syncKind);
 				theDocument.addDocumentListener(listener);
 				LanguageServerWrapper.this.connectedDocuments.put(uri, listener);
-				return listener.lastChangeFuture();
+				return listener.didOpenFuture();
```

Comment:
It's a pedantic point, but I think it better captures the original behaviour and also what I thought the intent was in terms of what `connect()` should be returning - i.e. a hook to allow async requests to be sent to the language server at (or the ability to block until) the point when the server is guaranteed to have been told about the document being opened.

It also goes to the heart of what we're trying to prevent with this change: client code trying to achieve a sequential ordering of interactions with the language server for a particular document, but doing it incorrectly.

If client code cares about ordering then it shouldn't interact with `lastChangeFuture` directly because that doesn't really offer any guarantees.

Possible sequence:

Document modified, with change 'A' sent to the server.

Client retrieves `lastChangeFuture` at time B.

Client async composes operation C on this future.

On another thread (or even on the same thread) something retrieves `lastChangeFuture`, gets the same object as at time B, and async composes operation D on it.

The only thing that is guaranteed is that the server will see change A before C or D is sent. Nothing is guaranteed about whether C or D is sent first, nor about which operation returns first.

Using `lastChangeFuture` doesn't really give any guarantees, other than that the document is (possibly) in a slightly later state than it would have been at `didOpenFuture`, which isn't really any use. If you care about sequencing then you need an API like the one we propose which offers 'after A, but before anything else', if not then just use `didOpenFuture` which just promises that the document has been loaded by the server.

Does that make any kind of sense?

Reply:
> The only thing that is guaranteed is that the server will see change A before C or D is sent

Right, but isn't it a good thing to guarantee? If I understand correctly this particular part of the proposal (using didOpen instead of lastChange) does remove this guarantee. I fail at seeing how it can improve some cases. Can you please provide insights about what is better with this?

Reply:
Well, currently I think it is only invoked by `LanguageServerWrapper.connect(...)` which in turn is just used to ensure the document has been connected. If so, using `didOpen` allows this guarantee to be met earlier.

If client code needs a 'happens after' guarantee then `executeOnCurrentVersionAsync` also does that, but offers the much stronger ability to ensure point-in-time consistency (or rather, to detect point-in-time inconsistency!). Since we're not keen on exposing more API than we have to I thought it made sense to encourage controlled access via the new method.

I guess what exposing `lastChange` would do is allow someone to schedule a potentially long-running call without forcing change notifications to queue behind it, but as noted above nothing was actually using that.

What do you think? I guess I don't feel that strongly about it, so it's down to what we think needs exposing in an API...